src/contracts/solana/ChainSignaturesContract.ts

```diff
@@ -0,0 +1,345 @@
+    const [pda] = PublicKey.findProgramAddressSync(
+      [Buffer.from('program-state')],
+      this.programId
+    )
+    return pda
+  }
+
+  async getDerivedPublicKey(args: {
+    path: string
+    predecessor: string
+  }): Promise<UncompressedPubKeySEC1> {
+    const pubKey = cryptography.deriveChildPublicKey(
+      await this.getPublicKey(),
+      args.predecessor,
+      args.path,
+      KDF_CHAIN_IDS.SOLANA
+    )
+
+    return pubKey
+  }
+
+  async getPublicKey(): Promise<UncompressedPubKeySEC1> {
+    return najToUncompressedPubKeySEC1(this.rootPublicKey)
+  }
+
+  /**
+   * Sends a transaction to the program to request a signature, then
+   * polls for the signature result. If the signature is not found within the retry
+   * parameters, it will throw an error.
+   */
+  async sign(
+    args: SignArgs & { feePayer?: Wallet },
+    options: SignOptions = {
+      sign: {
+        algo: '',
+        dest: '',
+        params: '',
+      },
+      retry: {
+        delay: 5000,
+        retryCount: 12,
+      },
+    }
+  ): Promise<RSVSignature> {
+    const wallet = this.provider.wallet
+
+    const feePayer = args.feePayer || wallet
+
+    const requestId = this.getRequestId(args, options.sign)
+
+    const eventPromise = this.listenForSignatureEvents({
+      requestId,
+      payload: args.payload,
+      path: args.path,
+      options: options.retry,
+    })
+
+    const hash = await this.program.methods
+      .sign(
+        Array.from(args.payload),
+        args.key_version,
+        args.path,
+        options.sign?.algo || '',
+        options.sign?.dest || '',
+        options.sign?.params || ''
+      )
+      .accounts({
+        requester: wallet.publicKey,
+        feePayer: feePayer.publicKey,
+      })
+      .rpc()
+
+    try {
+      const pollResult = await eventPromise
+
+      if (!pollResult) {
+        throw new SignatureNotFoundError(requestId, { hash })
+      }
+
+      if ('error' in pollResult) {
+        throw new SignatureContractError(pollResult.error, requestId, { hash })
+      }
+
+      return pollResult as RSVSignature
+    } catch (error) {
+      if (
+        error instanceof SignatureNotFoundError ||
+        error instanceof SignatureContractError
+      ) {
+        throw error
+      } else {
+        throw new SigningError(
+          requestId,
+          { hash },
+          error instanceof Error ? error : undefined
+        )
+      }
+    }
+  }
+
+  /**
+   * Listens for signature or error events matching the given requestId.
+   * Sets up listeners for both event types and returns a promise that resolves when
+   * either a valid signature or an error is received.
+   */
+  async listenForSignatureEvents({
+    requestId,
+    payload,
+    path,
+    options,
+  }: {
+    requestId: string
+    payload: number[]
+    path: string
+    options?: RetryOptions
+  }): Promise<RSVSignature | SignatureErrorData | undefined> {
+    const delay = options?.delay ?? 5000
```

Comment:
We are defining defaults for the second time here.

Reply:
the fallback values are only defined once the only other approach is to have a fallback object for `options` and I prefer this to that syntactically